### DISCUSSION TITLE
posts/pawnyable-0201/

### DISCUSSION THREAD

**Discussion by giscus** [brant-ruan/fernweh-comments#17] [Announcements]
# posts/pawnyable-0201/

Linux Kernel PWN | 040201 Pawnyable之栈溢出

https://blog.wohin.me/posts/pawnyable-0201/

**Comment by X1NRI** [comment]
这个gadget确实也卡了我很长时间，感谢教程！

**Reply by brant-ruan** [reply]
不客气，加油！

**Reply by ze-mu-zhou** [reply]
+1，感谢博主的教程，学到了很多，没有博主的探索我可能需要花费更长的时间才能明白这一点

**Reply by brant-ruan** [reply]
哈哈，一起加油！

**Comment by lionheartys** [comment]
在ret2usr的exp中，我在自己的实验环境里尝试博主的exp会出现提权成功但是system("/bin/sh") segment fault的问题，我调试了下发现是system函数内部逻辑中某个内核函数的出现错误。后来换成execve来执行就可以成功get shell，想问下博主有没有遇到过同样的问题

**Reply by brant-ruan** [reply]
不记得了，不过印象里几年前好像有看到过讨论system弹shell和execve的区别，以及system弹shell可能会失败。有点年头了，你可以调试一下看看。

**Reply by lionheartys** [reply]
OK，谢谢博主

**Reply by ze-mu-zhou** [reply]
根本原因是因为栈对齐
iretq 返回用户态后，rsp 可能不满足 x86-64 ABI 要求的 16 字节对齐
glibc 内部（尤其是 do_system）会用到大量 SSE 指令比如 movaps ，栈未对齐会导致直接 segfault 
而如果我们直接使用 execve 是单次 syscall，直接替换整个进程映像，不受影响

次要原因是因为
system() = fork() + execve() + waitpid() + 信号处理
可以看出 system 内部流程复杂，依赖的进程状态远比 execve() 多
所以在内核利用中我们尽量使用简单容易调试的 execve()
